#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [3]:
import pandas as pd

In [5]:
df = pd.read_csv('../data/supermarket_sales.csv')

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [5]:
df.groupby('Branch')['Total'].sum().sort_values(ascending=False)

Branch
C    110568.7065
A    106200.3705
B    106197.6720
Name: Total, dtype: float64

Філіал С має найвищий сумарний прибуток

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [7]:
df.groupby(['Branch', 'City'])['Total'].agg(['max', 'sum'])

max          sum
Branch City                           
A      Yangon     1039.29  106200.3705
B      Mandalay   1022.49  106197.6720
C      Naypyitaw  1042.65  110568.7065

In [9]:
df.groupby(['Branch', 'City'])['Total'].sum().sort_values(ascending=False)

Branch  City     
C       Naypyitaw    110568.7065
A       Yangon       106200.3705
B       Mandalay     106197.6720
Name: Total, dtype: float64

В місті Naypyitaw знаходиться філіал С з найвищим прибуком

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [11]:
invoices_by_category = df.pivot_table(index='Branch', columns='Product line', values='Invoice ID', aggfunc='count', fill_value=0)

print(invoices_by_category)

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45  


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [15]:
branch_with_max_invoices = invoices_by_category['Electronic accessories'].idxmax()
print(f"Філіал з найбільшою кількістю інвойсів для категорії 'Electronic accessories': {branch_with_max_invoices}")

Філіал з найбільшою кількістю інвойсів для категорії 'Electronic accessories': A


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


Гіпотеза 1:
Питання: Які лінії продуктів (Product line) найбільше купують залежно від міста (City)?

Аналітика: Це допоможе зрозуміти, які продукти популярні в різних містах, що дозволить краще управляти запасами та адаптувати маркетингові стратегії під конкретні регіони.

In [25]:
df.groupby(['City', 'Product line']).size().unstack()

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
City                                                                            
Mandalay                          55                   62                  50   
Naypyitaw                         55                   65                  66   
Yangon                            60                   51                  58   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
City                                                                    
Mandalay                     53                  50                 62  
Naypyitaw                    52                  45                 45  
Yangon                       47                  65                 59

Рекомендації: 
1.Відповідно до даних, бізнес може адаптувати запаси під кожне місто. Наприклад, у Mandalay можна збільшити запаси "Fashion accessories" і "Sports and travel", оскільки ці категорії мають високий попит. У Naypyitaw варто зосередитися на "Food and beverages", а в Yangon — на "Home and lifestyle" і "Electronic accessories". Це зменшить ризик надлишку товарів, які менш популярні в певних містах. 
2.Бізнес може розробити спеціальні маркетингові кампанії, орієнтовані на популярні продукти в кожному місті. Наприклад, у Yangon варто більше інвестувати в рекламу "Home and lifestyle", тоді як у Naypyitaw можна активно просувати "Food and beverages". Це допоможе залучити більше клієнтів і збільшити продажі. 
3.Міста з різними уподобаннями можуть стати тестовими майданчиками для нових продуктів. Наприклад, якщо у Mandalay добре продаються "Fashion accessories", нові продукти цієї категорії можна перш за все запускати саме там.

Гіпотеза 2:
Питання: Як податки впливають на загальну рентабельність?
Аналітика: Це дозволить оптимізувати витрати та податкові зобов'язання для покращення загальної фінансової ефективності.

In [50]:
total_tax = df['Tax 5%'].sum()
total_cogs = df['cogs'].sum()

print(f"Total Tax: {total_tax}")
print(f"Total COGS: {total_cogs}")

if total_cogs > 0:
    tax_cogs = total_tax / total_cogs
    print(f"Tax to COGS Ratio: {tax_cogs:.2%}")

Total Tax: 15379.368999999999
Total COGS: 307587.38
Tax to COGS Ratio: 5.00%


Рекомендації: Сума податків (15,379.37) є досить суттєвою, але для повного розуміння необхідно порівняти ці дані з загальними доходами компанії.
Витрати (COGS): Загальна сума витрат на виробництво та закупівлю товарів складає 307,587.38. Це значна сума, яка впливає на рентабельність компанії.
Співвідношення податків до витрат (5.00%): Це показує, що податки складають 5% від загальних витрат на товари. 
1. Варто перевірити, чи є можливість оптимізації податкових зобов'язань. Можливо, варто проконсультуватися з податковим консультантом, щоб дізнатися про доступні пільги або способи зменшення податкового навантаження.
2. Варто преглянути витрати на закупівлю товарів, щоб зрозуміти, чи можна їх зменшити. Це може включати переговори з постачальниками для отримання кращих умов.
3. Можна порівняти співвідношення податків до витрат з галузевими стандартами або середніми показниками. Це дозволить зрозуміти, чи співвідношення є нормальним, чи є місце для покращення.

Гіпотеза 3: Аналіз рентабельності за типами клієнтів. Якщо члени приносать більше прибутку, подумайте про розробку програми лояльності або спеціальних акцій для них.

In [66]:
df.groupby('Customer type').agg(total_sales = ('Total', 'sum'), avg_gross_margin_percentage = ('gross margin percentage', 'mean'))

total_sales  avg_gross_margin_percentage
Customer type                                          
Member          164223.444                     4.761905
Normal          158743.305                     4.761905

Результат:Загальний дохід від членів трохи вищий за дохід від звичайних клієнтів.
Середня рентабельність (gross margin percentage) є однаковою для обох типів клієнтів — 4.76%.
Висновок: Незалежно від типу клієнта, рентабельність залишається стабільною. 

Рекомендації:
1. Зосередьтеся на залученні більше членів: оскільки доходи від членів є вищими, розгляньте можливість розширення програми лояльності або створення нових переваг для членів. Це може включати спеціальні акції, знижки або ексклюзивні продукти для членів.
2. Розгляньте можливість покращення програми лояльності: оскільки рентабельність однакова для обох типів клієнтів, підвищення кількості членів може бути вигідним для бізнесу. Можливо, варто інвестувати в маркетинг та рекламу для залучення нових членів.
